source/simulator/nullspace.cc

```diff
@@ -81,135 +90,184 @@ namespace aspect
+  template <int dim>
+  void Simulator<dim>::remove_net_linear_momentum( LinearAlgebra::BlockVector &relevant_dst,
+                                                   LinearAlgebra::BlockVector &tmp_distributed_stokes,
+                                                   bool use_constant_density)
+  {
+    Assert(introspection.block_indices.velocities != introspection.block_indices.pressure,
+           ExcNotImplemented());
 
-    if (parameters.nullspace_removal & NullspaceRemoval::net_translation_y)
-      funcs.push_back(std_cxx1x::shared_ptr<TensorFunction<1,dim> >(new internal::Translation<dim>(1)));
+    // compute and remove net linear momentum from velocity field, by computing
+    // \int \rho (v + v_const) = 0
 
-    if (parameters.nullspace_removal & NullspaceRemoval::net_translation_z)
-      {
-        //Only do z direction if dim == 3
-        AssertThrow( dim == 3, ExcMessage("Can't remove z translational mode in 2 dimensions"));
-        funcs.push_back(std_cxx1x::shared_ptr<TensorFunction<1,dim> >(new internal::Translation<dim>(2)));
-      }
+    //vector for storing the correction to the velocity field
+    LinearAlgebra::Vector correction(tmp_distributed_stokes.block(introspection.block_indices.velocities));
 
-    if (funcs.size()>0)
-      {
-        Assert(introspection.block_indices.velocities != introspection.block_indices.pressure,
-               ExcNotImplemented());
-
-        net_rotations_translations.resize(funcs.size());
-        for (unsigned int i=0; i<funcs.size(); ++i)
-          {
-            // for each of the null space dimensions, set up
-            // a vector, fill it with an element of the null space,
-            // and normalize it
-            net_rotations_translations[i].reinit(
-              introspection.index_sets.system_partitioning[introspection.block_indices.velocities],
-              mpi_communicator);
-            interpolate_onto_velocity_system(*funcs[i],
-                                             net_rotations_translations[i]);
-            net_rotations_translations[i] /= net_rotations_translations[i].l2_norm();
-          }
+    QGauss<dim> quadrature(parameters.stokes_velocity_degree+1);
+    FEValues<dim> fe(mapping, finite_element, quadrature,
+                     UpdateFlags(update_quadrature_points | update_JxW_values | update_values));
 
-      }
-  }
+    typename DoFHandler<dim>::active_cell_iterator cell;
+    std::vector<Point<dim> > q_points(quadrature.size());
+    std::vector<Vector<double> > fe_vals(quadrature.size(), Vector<double>(finite_element.n_components()));
 
+    Tensor<1,dim> local_momentum;
+    double local_mass;
 
+    //loop over all local cells
+    for (cell = dof_handler.begin_active(); cell != dof_handler.end(); ++cell)
+      if (cell->is_locally_owned())
+        {
+          fe.reinit (cell);
+          q_points = fe.get_quadrature_points();
```

Comment:
You're copying the whole array here. You can avoid this by declaring `q_points` as a reference type right here:
```
  const std::vector<Point<dim> > &q_points = fe.get_quadrature_points();
```